### PR TITLE
gearpump-34 update developer document.

### PR DESCRIPTION
**Pull request by whjiang**

### CODE REVIEW

**Comment by karol-brejna-i** [comment]
Small thing:

IDE setup link (beginning of the page) doesn't work. It points to:
https://github.com/whjiang/incubator-gearpump/blob/gearpump-34/CONTRIBUTING.md#IDE-setup

The corresponding section id is: 
https://github.com/whjiang/incubator-gearpump/blob/gearpump-34/CONTRIBUTING.md#ide-setup

(upper case vs lower case letters).

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,41 +1,125 @@
...
+
+1. You need to create a working branch
 
   ```
   git branch branch_issueId
-  git checkout branch_issusId
+  git checkout branch_issueId
   ```
 
   Work on the branch, make changes, then push to your forked gearpump repo http://github.com/<git-user_id>/gearpump
```
Comment (karol-brejna-i) [inline CONTRIBUTING.md]: ... forked Gearpump repo... instead of *g*earpump rebpo

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -53,40 +137,242 @@ No work should ever be done in the forked master. Another way to do this is to
...
+Committer will review your code periodically. When there is any comment/feedback from committer(s), it's contributor's duty to update the pull request correspondingly. 
+
+When the merge was done by committer, you can optionally [delete your PR branch](https://help.github.com/articles/deleting-unused-branches).
+
+
+<a name="build-and-test"></a>
+
+# Build and test
+Though without a development environment setup, you can still contribute to Gearpump by reporting ideas, documentations, bug reports and feature requests.
+It is highly recommended to set up a development emnvironment to make any code contribution.
```
Comment (karol-brejna-i) [inline CONTRIBUTING.md]: environment

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -53,40 +137,242 @@ No work should ever be done in the forked master. Another way to do this is to
 
    You can also use this when you want to squash(merge) multiple commits into one.
    ```git rebase -i``` will popup a window, which allow you to squash(merge) multiple commits into one commit.
```
Comment (karol-brejna-i) [inline CONTRIBUTING.md]: 'popup' should probably be 'pop up'...

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,41 +1,125 @@
-This documents are for committers which have direct write permission to intel-hadoop/gearpump
+# Developer documentation
 
-Commit Guideline
-========================
-1.	For all commit(except doc), must create an issue id.
```
Comment (karol-brejna-i) [inline CONTRIBUTING.md]: for every?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -53,40 +137,242 @@ No work should ever be done in the forked master. Another way to do this is to
...
+<a name="build-and-test"></a>
+
+# Build and test
+Though without a development environment setup, you can still contribute to Gearpump by reporting ideas, documentations, bug reports and feature requests.
+It is highly recommended to set up a development emnvironment to make any code contribution.
+
+<a name="local-copy"></a>
+## Clone Gearpump repository and make a local copy
+If you just want to study Gearpump source code, it is optional to perform following steps. 
+But, if you plans to contribute to Gearpump's code base, it is necessary to perform following steps:
```
Comment (karol-brejna-i) [inline CONTRIBUTING.md]: plan instead of plans

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -53,40 +137,242 @@ No work should ever be done in the forked master. Another way to do this is to
...
+To run unit test, you can run
+```bash
+  sbt test
+```
+
+Gearpump has an integration test system which is based on Docker. Please check [the instructions](integrationtest/README.md).
+
+<a name="build-doc"></a>
+## How to build Gearpump documentation
+To build Gearpump document, use
```
Comment (karol-brejna-i) [inline CONTRIBUTING.md]: I think comma is unnecessary here.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -53,40 +137,242 @@ No work should ever be done in the forked master. Another way to do this is to
...
+It is highly recommended to perform [package build](#build) before IDE setup.
+
+<a name="code-style"></a>
+## Gearpump code style
+
+Gearpump follows the standard Scala style, just like other Scala projects. This style is:
+
+1. Mainly, we follow the [Spark Scala style](https://github.com/databricks/scala-style-guide).
+
+2. We allows some exceptions: e.g. allowing using !, ? for Akka to send actor message.
```
Comment (karol-brejna-i) [inline CONTRIBUTING.md]: we allow

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -53,40 +137,242 @@ No work should ever be done in the forked master. Another way to do this is to
...
+
+<a name="committer-work"></a>
+# Committer section
+_This section applies to committers only._
+
+<a name="approve-pull-request"></a>
+## Approve pull request
+It's committer's duty to review pull requests from contributors. 
+
+Any PR ready to merge shall have at least one +1(s) and no -1(s). And any merge shall wait another 24 hours after the first +1 received to wait for potential comments.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: other than the one who authored the patch

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,41 +1,125 @@
...
+This section explains how to make a contribution. 
+
+
+<a name="report-bug"></a>
+
+## Report a bug
+
+To report a bug you should [open an issue](https://issues.apache.org/jira/browse/GEARPUMP) in our issue tracker that
+summarizes the bug.  Set the form field "Issue type" to "Bug".  If you have not used the issue tracker before you will
+need to register an account (free), log in, and then click on the blue "Create Issue" button in the top navigation bar.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: red "Create Issue" button ?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,41 +1,125 @@
...
+with.
+
+
+<a name="request-feature"></a>
+
+## Request a new feature
 
-4. Create a working branch
+To request a new feature you should [open an issue](https://issues.apache.org/jira/browse/GEARPUMP) in our issue tracker
+and summarize the desired functionality.  Set the form field "Issue type" to "New feature".  If you have not used the
```
Comment (manuzhang) [inline CONTRIBUTING.md]: What about "Improvement" and other issue types ? Any guide for "Priority" ?

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,41 +1,125 @@
...
+Before working on code contribution, you need to prepare your [development environment](#build-and-test).
+
+To work on a code contribution, following process is suggested:
+
+1. You need to create a working branch
 
   ```
   git branch branch_issueId
-  git checkout branch_issusId
+  git checkout branch_issueId
```
Comment (manuzhang) [inline CONTRIBUTING.md]: make sure we are on latest master branch beforehand and "git checkout -b branch_issueId" will do.

File: CONTRIBUTING.md (hunk 1 above)
Comment (manuzhang) [inline CONTRIBUTING.md]: `<git-user_id>` is not rendered properly maybe `${git-user_id}` ?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -44,7 +128,7 @@ No work should ever be done in the forked master. Another way to do this is to
   git push origin branch_issueId
   ```
 
-5. When there is changes in upstream/master, rebase your work on upstream/master. with
+2. When there is changes in upstream/master, rebase your work on upstream/master.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: `there are` instead of `there is`

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -53,40 +137,242 @@ No work should ever be done in the forked master. Another way to do this is to
...
   After the code is successfully rebased, a window will pop up to edit the commit log, edit it then save and exit.
-7. After rebase, now you have a clean log history. push to your remote working branch
+4. After rebase, now you have a clean log history. push to your remote working branch
 
   ```
   git push origin branch_issueId
   ```
 
-  If commits have already been pushed to <git-user-id>/gearpump fork on github, you will have to "git push --force" to overwrite them with squashed commits.
+  If commits have already been pushed to your <git-user-id>/incubator-gearpump fork on github, you will have to "git push --force" to overwrite them with squashed commits.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: `GitHub` instead of `github`

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -53,40 +137,242 @@ No work should ever be done in the forked master. Another way to do this is to
...
+  cd incubator-gearpump
+  ```
+
+3. Add apache/incubator-gearpump as an external repo 'upstream' by following the [guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/).
+
+  ```bash
+  git remote add upstream https://github.com/apache/incubator-gearpump.git
+  ```
+
+4. In local master branch, periodically sync the forked master with the main master with
```
Comment (manuzhang) [inline CONTRIBUTING.md]: upstream master instead of main master is less confusing

File: CONTRIBUTING.md (hunk 7 above)
Comment (manuzhang) [inline CONTRIBUTING.md]: we'd better make a list of what those exceptions are

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -53,40 +137,242 @@ No work should ever be done in the forked master. Another way to do this is to
...
+
+<a name="merge-pull-request"></a>
+## Merge pull request
+For each committer, it is suggested to make a local clone of Apache Gearpump repository and make it solely for PR merge purpose. 
+```bash
+git clone https://git-wip-us.apache.org/repos/asf/incubator-gearpump
+
+```
+**NOTE: Only Apache Gearpump repository is writable. The github mirror is READONLY.
+
```
Comment (manuzhang) [inline CONTRIBUTING.md]: missing trailing "**"

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -53,40 +137,242 @@ No work should ever be done in the forked master. Another way to do this is to
...
+
+<a name="merge-pull-request"></a>
+## Merge pull request
+For each committer, it is suggested to make a local clone of Apache Gearpump repository and make it solely for PR merge purpose. 
+```bash
+git clone https://git-wip-us.apache.org/repos/asf/incubator-gearpump
+
+```
+**NOTE: Only Apache Gearpump repository is writable. The github mirror is READONLY.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: missing trailing "**"

**Comment by manuzhang** [comment]
"github", "GitHub" and "github.com" are used in several places to referred to the website. Better consistently use "GitHub".

File: CONTRIBUTING.md (hunk 1 above)
Comment (whjiang) [inline CONTRIBUTING.md]: I am not sure about this one. I inherit it from the previous documenet written by Sean. 

@clockfly can you verify the variable?

File: CONTRIBUTING.md (hunk 4 above)
Comment (whjiang) [inline CONTRIBUTING.md]: Yes. I think this is the guideline from beginning.

File: CONTRIBUTING.md (hunk 7 above)
Comment (whjiang) [inline CONTRIBUTING.md]: yes. can anyone give me a full list of exceptions?

File: CONTRIBUTING.md (hunk 10 above)
Comment (whjiang) [inline CONTRIBUTING.md]: these sections are borrowed from Storm guide. I am not sure whether we want to have all these assignable by contributor or we shall adjust it later by committers?

And I prefer to let committer adjust the "Priority" field instead of committer.

**Comment by whjiang** [comment]
updated according to your comments, please review again.

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -1,92 +1,378 @@
...
+with.
+
+
+<a name="request-feature"></a>
+
+## Request a new feature
+
+To request a new feature you should [open an issue](https://issues.apache.org/jira/browse/GEARPUMP) in our issue tracker
+and summarize the desired functionality.  Set the form field "Issue type" to "New feature".  If you have not used the
+issue tracker before you will need to register an account (free), log in, and then click on the blue "Create Issue"
```
Comment (manuzhang) [inline CONTRIBUTING.md]: red button ?

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -1,92 +1,378 @@
...
+button in the top navigation bar.
 
-4. Create a working branch
+
+<a name="contribute-code"></a>
+
+## Contribute code/document by creating Pull Request
+Before you set out to contribute code we recommend that you familiarize yourself with the Gearpump codebase and corresponding development document at Gearpump website.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: link to Gearpump website ?

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -1,92 +1,378 @@
...
+    2. Comment on the issue with your plan for implementing the issue.  Explain what pieces of the codebase you are
+       going to touch and how everything is going to fit together.
+    3. Gearpump committers will iterate with you on the design to make sure you are on the right track.
+    4. Implement your issue, create a pull request (see below), and iterate from there.
+
+
+### Contribution Guideline
+1.	For all commits, must create an issue id.
+2.	For all commit log messages, must contain the issue id, Like this: fix #issueId, comments.
+3.	For all PRs (pull request), the title must contains issue Id.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: =>  must contain issue id

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -1,92 +1,378 @@
...
+1. You need to create a working branch (before that, please make sure your local master is synced with **upstream** master.)
 
   ```
   git branch branch_issueId
-  git checkout branch_issusId
+  git checkout branch_issueId
   ```
 
-  Work on the branch, make changes, then push to your forked gearpump repo http://github.com/<git-user_id>/gearpump
+  Work on the branch, make changes, then push to your forked Gearpump repo http://github.com/<git-user_id>/incubator-gearpump.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: "http://github.com" is rendered into a link while the rest into plain texts. Better suppress the link

File: CONTRIBUTING.md [hunk 21]
```diff
@@ -1,92 +1,378 @@
...
 
   ```
   git push origin branch_issueId
   ```
 
-5. When there is changes in upstream/master, rebase your work on upstream/master. with
+2. When there are changes in upstream/master, rebase your work on upstream/master.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: we need to fetch upstream changes before rebase

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -1,92 +1,378 @@
...
 
    ```bash
    git checkout branch_issueId
    git rebase -i upstream/master
    ```
 
    You can also use this when you want to squash(merge) multiple commits into one.
-   ```git rebase -i``` will popup a window, which allow you to squash(merge) multiple commits into one commit.
-   For example I might have 12 commits in my branch. "rebase -i upstream/master" opens a nice editor where you can mark some commits to be squashed(merged) into prior commits, and make 1 big commit (or several) out of it. In this way, I can tidy up what will be committed to the project master's history since otherwise my commit messages are like "not working" or "got it working" or "more fix" or "merged <git-user-id>/gearpump to master".
+   ```git rebase -i``` will pop up a window, which allow you to squash(merge) multiple commits into one commit.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: allow => allows

File: CONTRIBUTING.md [hunk 23]
```diff
@@ -1,92 +1,378 @@
...
+   For example I might have 12 commits in my branch. ```git rebase -i upstream/master``` opens a nice editor where you can mark some commits to be squashed(merged) into prior commits, and make 1 big commit (or several) out of it. In this way, I can tidy up what will be committed to the project master's history since otherwise my commit messages are like "not working" or "got it working" or "more fix" or "merged <git-user-id>/gearpump to master".
```
Comment (manuzhang) [inline CONTRIBUTING.md]: `<git-user-id>` is not rendered

File: CONTRIBUTING.md [hunk 24]
```diff
@@ -1,92 +1,378 @@
...
   After the code is successfully rebased, a window will pop up to edit the commit log, edit it then save and exit.
-7. After rebase, now you have a clean log history. push to your remote working branch
+4. After rebase, now you have a clean log history. push to your remote working branch
 
   ```
   git push origin branch_issueId
   ```
 
-  If commits have already been pushed to <git-user-id>/gearpump fork on github, you will have to "git push --force" to overwrite them with squashed commits.
+  If commits have already been pushed to your <git-user-id>/incubator-gearpump fork on GitHub, you will have to "git push --force" to overwrite them with squashed commits.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: * `<git-user-id>` is not rendered
* `<git-user-id>` => `<git-user_id>`

File: CONTRIBUTING.md [hunk 25]
```diff
@@ -1,92 +1,378 @@
...
-6. Open a PR, which is a one-click thing in github.com; it knows you likely are opening a PR against upstream master. [Guide](https://help.github.com/articles/creating-a-pull-request) is here.
-7. [Merge PR](https://help.github.com/articles/merging-a-pull-request), [delete branch](https://help.github.com/articles/deleting-unused-branches).
+5. Ensure all the unit tests and integration tests are passed, check [Test](#test) for details.
+6. Open a Pull Request, which is a one-click thing in github.com; it knows you likely are opening a PR against upstream master. [Guide](https://help.github.com/articles/creating-a-pull-request) is here.
+
+<a name="code-review"></a>
+# Code Review
+Committer will review your code periodically. When there is any comment/feedback from committer(s), it's contributor's duty to update the pull request correspondingly. 
+
+When the merge was done by committer, you can optionally [delete your PR branch](https://help.github.com/articles/deleting-unused-branches).
```
Comment (manuzhang) [inline CONTRIBUTING.md]: was => is

File: CONTRIBUTING.md [hunk 26]
```diff
@@ -1,92 +1,378 @@
...
+
+## Report a bug
+
+To report a bug you should [open an issue](https://issues.apache.org/jira/browse/GEARPUMP) in our issue tracker that
+summarizes the bug.  Set the form field "Issue type" to "Bug".  If you have not used the issue tracker before you will
+need to register an account (free), log in, and then click on the red "Create Issue" button in the top navigation bar.
+
+In order to help us understand and fix the bug it would be great if you could provide us with:
+
+1. The steps to reproduce the bug.  This includes information about e.g. the Gearpump version you were using, the deployment model, etc.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: were  => are

File: CONTRIBUTING.md [hunk 27]
```diff
@@ -1,92 +1,378 @@
...
+# Code Review
+Committer will review your code periodically. When there is any comment/feedback from committer(s), it's contributor's duty to update the pull request correspondingly. 
+
+When the merge was done by committer, you can optionally [delete your PR branch](https://help.github.com/articles/deleting-unused-branches).
+
+
+<a name="build-and-test"></a>
+
+# Build and test
+Though without a development environment setup, you can still contribute to Gearpump by reporting ideas, documentations, bug reports and feature requests.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: just "bugs" instead of "bug reports" will do

File: CONTRIBUTING.md [hunk 28]
```diff
@@ -1,92 +1,378 @@
...
+It is highly recommended to set up a development environment to make any code contribution.
+
+<a name="local-copy"></a>
+## Clone Gearpump repository and make a local copy
+If you just want to study Gearpump source code, it is optional to perform following steps. 
+But, if you plan to contribute to Gearpump's code base, it is necessary to perform following steps:
+
+1. [Fork](https://help.github.com/articles/fork-a-repo) in GitHub from https://github.com/apache/incubator-gearpump to create a /incubator-gearpump repo. 
+After fork, you will have a new repo at https://github.com/<git-user_id>/incubator-gearpump.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: `<git-user_id>` is not rendered

File: CONTRIBUTING.md [hunk 29]
```diff
@@ -1,92 +1,378 @@
...
+To make a compilation of Gearpump, you can execute:
+```bash
+  sbt compile pack
+```
+
+To build a Gearpump package, you can execute following commands:
+
+```bash
+  ## The target package path: output/target/gearpump-${version}.zip
+  sbt clean +assembly +packArchiveZip
```
Comment (manuzhang) [inline CONTRIBUTING.md]: do we prefer `packArchive` to generate ".tar.gz" ?

File: CONTRIBUTING.md [hunk 30]
```diff
@@ -1,92 +1,378 @@
...
+To build Gearpump document use
+```bash
+   ## the 2nd argument indicates whether to build API doc or not
+   ## for release, we need to use '1' to build it (time consuming)
+   docs/build_doc.sh  2.11  0 
+```  
+
+<a name="ide-setup"></a>
+## IDE setup
+IDE environment can be set up on either Windows, Linux and MAC platform. You can choose the one you prefer.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: Mac, I think

File: CONTRIBUTING.md [hunk 31]
```diff
@@ -1,92 +1,378 @@
...
+
+<a name="write-integrationtest"></a>
+## How to write integration test
+TBD
+
+<a name="write-doc"></a>
+## How to write document
+Documentation contributions are very welcome!
+
+You can contribute documentation by pull request, as same as code contribution.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: "as same as" => same as

File: CONTRIBUTING.md [hunk 32]
```diff
@@ -1,92 +1,378 @@
...
+Main directory is ```docs/```, and you can refer to docs/README.md for how to build / test documentation.
+
+NOTE: this documentation is the release documentation for Gearpump, not the website documentation at Apache website, which was maintained by another repository.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: was maintained => is maintained

File: CONTRIBUTING.md (hunk 18 above)
Comment (whjiang) [inline CONTRIBUTING.md]: what is the official Gearpump website now?
gearpump.io or the apache one?

File: CONTRIBUTING.md (hunk 24 above)
Comment (whjiang) [inline CONTRIBUTING.md]: what is "backtick"?

Comment (manuzhang) [inline CONTRIBUTING.md]: oh, this guy "`"

File: CONTRIBUTING.md (hunk 29 above)
Comment (whjiang) [inline CONTRIBUTING.md]: I am not sure here. It was inherited from old version.

File: CONTRIBUTING.md (hunk 18 above)
Comment (manuzhang) [inline CONTRIBUTING.md]: apache website is not setup yet. let's use gearpump.io for now and update later.

**Comment by whjiang** [comment]
I updated the doc according to all the comments above. Only one thing I don't know how to handle: the <git-user_id> field. Any suggestion here? Using a common name like "Bob" here?

**Comment by manuzhang** [comment]
which field ?

**Comment by whjiang** [comment]
@manuzhang get-user_id

**Comment by manuzhang** [comment]
I think "fork" will do and GitHub users can understand. We don't need the address.

**Comment by whjiang** [comment]
updated.

File: CONTRIBUTING.md [hunk 33]
```diff
@@ -1,91 +1,379 @@
...
+```bash
+  ## The target package path: output/target/gearpump-${version}.zip
+  sbt clean +assembly +packArchiveZip
+```
+
+  After the build, there will be a package file gearpump-${version}.zip generated under output/target/ folder.
+
+  **NOTE:**
+The build requires network connection. If you are behind a proxy, make sure you have set the proxy in your env before running the build commands.
+For windows:
```
Comment (manuzhang) [inline CONTRIBUTING.md]: 1. windows => Windows
2. please add an empty line above

File: CONTRIBUTING.md [hunk 34]
```diff
@@ -1,91 +1,379 @@
...
+Though without a development environment setup, you can still contribute to Gearpump by reporting ideas, documentations, bugs and feature requests.
+It is highly recommended to set up a development environment to make any code contribution.
+
+<a name="local-copy"></a>
+## Clone Gearpump repository and make a local copy
+If you just want to study Gearpump source code, it is optional to perform following steps. 
+But, if you plan to contribute to Gearpump's code base, it is necessary to perform following steps:
+
+1. [Fork](https://help.github.com/articles/fork-a-repo) in GitHub from https://github.com/apache/incubator-gearpump to create a /incubator-gearpump repo. 
+After fork, you will have a new repo at https://github.com/your_username/incubator-gearpump.
```
Comment (manuzhang) [inline CONTRIBUTING.md]: a bit wordy, How about "[Fork](https://help.github.com/articles/fork-a-repo) `https://github.com/apache/incubator-gearpump` to `https://github.com/your_username/incubator-gearpump`"

**Comment by manuzhang** [comment]
a few more comments. +1 generally.

**Comment by whjiang** [comment]
updated according to comments.

**Comment by manuzhang** [comment]
please rebase your commit

File: CONTRIBUTING.md [hunk 35]
```diff
@@ -1,91 +1,379 @@
...
+  Gearpump JIRA ticket associated with the pull request (example: `GEARPUMP-1234`).
+
+        $ git checkout -b <local_test_branch>  # e.g. git checkout -b GEARPUMP-1234
+
+3. Merge the pull request into your local test branch.
+
+        $ git pull <remote_repo_url> <remote_branch>
+    You can use `./dev-tools/gearpump-merge.py <pull-number>` to produce the above command most of the time.
+
+4.  Assuming that the pull request merges without any conflicts:
```
Comment (manuzhang) [inline CONTRIBUTING.md]: I'm afraid an extra rebase step is needed to avoid an extra merge commit in case master is updated although having no conflicts with the branch.

**Comment by manuzhang** [comment]
After merging several commits, I think the merge script https://github.com/apache/spark/blob/master/dev/merge_spark_pr.py used by Spark works better and creates cleaner commit history (also close pr on rebase and cherry-pick).

**Comment by whjiang** [comment]
@manuzhang 
1. I have added the rebase step. 
2. Can you help to work on my branch to add the spark merge script and corresponding document update for how to use it? As you have used it, it will be more accurate for you to update the guide.

**Comment by manuzhang** [comment]
@whjiang sure. I'll work on it.

**Comment by manuzhang** [comment]
@whjiang updated. You may try out the new script.

**Comment by manuzhang** [comment]
Guys, any more comments ? If not, I'm going to merge this in.

**Comment by kkasravi** [comment]
+1
